homeassistant/components/yeelight/__init__.py

```diff
@@ -333,88 +337,147 @@ def async_get(cls, hass: HomeAssistant):
     def __init__(self, hass: HomeAssistant) -> None:
         """Initialize class."""
         self._hass = hass
-        self._seen = {}
         self._callbacks = {}
-        self._scan_task = None
+        self._host_discovered_events = {}
+        self._unique_id_capabilities = {}
+        self._host_capabilities = {}
+        self._track_interval = None
+        self._listener = None
+        self._connected_event = None
+
+    async def async_setup(self):
+        """Set up the scanner."""
+        if self._connected_event:
+            await self._connected_event.wait()
+            return
+        self._connected_event = asyncio.Event()
 
-    async def _async_scan(self):
-        _LOGGER.debug("Yeelight scanning")
-        # Run 3 times as packets can get lost
-        for _ in range(3):
-            devices = await self._hass.async_add_executor_job(discover_bulbs)
-            for device in devices:
-                unique_id = device["capabilities"]["id"]
-                if unique_id in self._seen:
-                    continue
-                host = device["ip"]
-                self._seen[unique_id] = host
-                _LOGGER.debug("Yeelight discovered at %s", host)
-                if unique_id in self._callbacks:
-                    self._hass.async_create_task(self._callbacks[unique_id](host))
-                    self._callbacks.pop(unique_id)
-                    if len(self._callbacks) == 0:
-                        self._async_stop_scan()
-
-        await asyncio.sleep(DISCOVERY_INTERVAL.total_seconds())
-        self._scan_task = self._hass.loop.create_task(self._async_scan())
+        async def _async_connected():
+            self._listener.async_search()
+            self._connected_event.set()
+
+        self._listener = SSDPListener(
```

Comment:
Ideally this becomes aware of the network integration in a future PR after the interfaces in https://github.com/home-assistant/core/pull/54832 are available